src/bm_sim/calculations.cpp

```diff
@@ -360,36 +360,110 @@ struct csum16 {
   }
 };
 
+// Concatenate the next n bytes and emit as the specified integer type.
+// Pad zeros at tail when given bytes are shorter than the integer type.
+// For example, emitAsInteger<uint32_t>(&{0xab, 0xcd}, 2) == 0xabcd0000.
+template <typename T>
+T emitAsInteger(const char *buf, size_t n) {
+  static constexpr size_t type_size = sizeof(T);
+  assert(n <= type_size);
+  T res = 0;
+  if (n > 0) {
+    for (size_t offset = 0; offset < n; offset++) {
+      res = (res << 8) | (static_cast<T>(buf[offset]) & 0xff);
+    }
+    res <<= ((type_size - n) * 8);
+  }
+  return res;
+}
+
 template <typename T>
 struct xor_hash {
   T operator()(const char *buf, size_t len) const {
     static constexpr size_t n = sizeof(T);
     T final_xor_value = 0;
-    T mask = 0xff;
     size_t byte = 0;
-    T t;
 
     /* Main loop - n bytes at a time */
     while (len >= n) {
-      t = 0;
-      for (size_t offset = 0; offset < n; offset++) {
-        t = (t << 8) + (static_cast<T>(buf[byte + offset]) & mask);
-      }
-      final_xor_value ^= t;
+      final_xor_value ^= emitAsInteger<T>(buf + byte, n);
       byte += n;
       len -= n;
     }
 
     /* Handle tail less than n bytes long */
-    t = 0;
-    for (size_t offset = 0; offset < len; offset++) {
-      t = (t << 8) + (static_cast<T>(buf[byte + offset]) & mask);
+    final_xor_value ^= emitAsInteger<T>(buf + byte, len);
+
+    return final_xor_value;
+  }
+};
+
+struct toeplitz {
+  using rss_key_t = RssMgr::rss_key_t;
+
+  toeplitz() {
+    update_key({0x00, 0x00, 0x00, 0x00});
+  }
+
+  uint32_t operator()(const char *buf, size_t len) const {
+    // Convert input bytes (in network byte order) to host words.
+    // Pad the last one with zeros if it's shorter than a word.
+    std::vector<uint32_t> input_words;
+    auto raw_input_words = reinterpret_cast<const uint32_t *>(buf);
+    const size_t n_words = len / 4;
+    const size_t n_bytes_remain = len % 4;
+    for (size_t i_word = 0; i_word < n_words; i_word++) {
+      input_words.push_back(ntohl(raw_input_words[i_word]));
```

Comment:
out-of-curiosity, does the dpdk implementation also require an `ntohl` conversion?

Reply:
In DPDK, the input is already in 4-bytes chunks, so they don't need the step to convert from raw bytes. See:

https://github.com/DPDK/dpdk/blob/72206323a5dd3182b13f61b25a64abdddfee595c/lib/hash/rte_thash.h#L177

Reply:
Also quoting the DPDK doc from https://doc.dpdk.org/guides/prog_guide/toeplitz_hash_lib.html:

> Both of above mentioned \_softrss\_ functions expect the tuple to be in “host” byte order and a multiple of 4 bytes in length.